src/components/chips/js/chipsController.js

```diff
@@ -577,23 +577,34 @@ MdChipsCtrl.prototype.hasMaxChipsReached = function() {
 
 /**
  * Updates the validity properties for the ngModel.
+ *
+ * TODO add the md-max-chips validator to this.ngModelCtrl.validators so that the validation will
+ * be performed automatically.
  */
 MdChipsCtrl.prototype.validateModel = function() {
   this.ngModelCtrl.$setValidity('md-max-chips', !this.hasMaxChipsReached());
   this.ngModelCtrl.$validate(); // rerun any registered validators
 };
 
-MdChipsCtrl.prototype.updateNgModel = function() {
-  this.ngModelCtrl.$setViewValue(this.items.slice());
-  // TODO add the md-max-chips validator to this.ngModelCtrl.validators so that
-  // the validation will be performed automatically on $viewValue change
-  this.validateModel();
+/**
+ * Function to handle updating the model, validation, and change notification when a chip
+ * is added, removed, or changed.
+ * @param {boolean=} skipValidation true to skip calling validateModel()
+ */
+MdChipsCtrl.prototype.updateNgModel = function(skipValidation) {
+  this.ngModelCtrl.$setDirty();
```

Comment:
Why not call `$setViewValue()`? This will break expected `ngModel` behavior.

Reply:
With what value? Calling `$setViewValue(this.items.slice())` is what caused these regressions (issues linked in OP). Calling `$setViewValue(this.items)` doesn't trigger `ngChange`, but I guess I could try that again now that we're doing it manually.

Reply:
`this.ngModelCtrl.$setViewValue(this.items);` does seem to work now with the manual triggering of `ngChange`. Will update.

Reply:
I see. So, yeah, calling `$setViewValue(this.items)` might not work (since it uses `===` to [compare to previous value][1]) and thus will not perform the expected operations (update classes, validate, call `$viewChangeListeners`).

As a workaround, maybe you could call `$viewChangeListeners` yourself, which I think it a better approach than manually calling `ngChange` ([relevant AngularJS code][2]):

```js
angular.forEach(ngModelCtrl.$viewChangeListeners, function(listener) {
  try {
    listener();
  } catch (e) {
    $exceptionHandler(e);
  }
});
```

[1]: https://github.com/angular/angular.js/blob/459997b482ef9597ce3505d16f7ebc3b9a14b9cb/src/ng/directive/ngModel.js#L713
[2]: https://github.com/angular/angular.js/blob/459997b482ef9597ce3505d16f7ebc3b9a14b9cb/src/ng/directive/ngModel.js#L781-L788

Reply:
Excellent, thank you. That seems to work and resolves the argument naming issue.